### PR TITLE
Feature/grid iterator

### PR DESCRIPTION
**Pull request by Jammy2211**
New type of grid, that given a function where the result improves by calculating it at high subgrid resolutions (e.g. a light profile caclualtion) the calculation uses iteratively higher resolutioon grids depending on the fractinal accuracy of the cacluation.

Grids are now implemented via a decorator which changes the behaviour of these functions depening on the grid that is passed. This is currently used for 3 grid types (normal sub grid, coordinates grid, iterator grid) and will most likely be extended to many grids in the future (interpolation, 3D grids).

This push is a work in progress to get an feedback on the implemention.

### CODE REVIEW

File: autoarray/structures/grids.py [hunk 1]
```diff
@@ -762,11 +764,188 @@ def blurring_grid_from_mask_and_kernel_shape(
                 grid_1d=blurring_grid_1d
             )
 
+    def structure_from_result(self, result):
```
Comment (rhayes777) [inline autoarray/structures/grids.py:767]: Type annotations would be nice

File: autoarray/structures/grids.py [hunk 2]
```diff
@@ -762,11 +764,188 @@ def blurring_grid_from_mask_and_kernel_shape(
...
+                 grid[2] = [0.75, -0.25]
+        |0|1|2|  grid[3] = [0.5,  -0.75]
+        |3|4|5|  grid[4] = [0.5,  -0.5]
+        |6|7|8|  grid[5] = [0.5,  -0.25]
+                 grid[6] = [0.25, -0.75]
+                 grid[7] = [0.25, -0.5]
+                 grid[8] = [0.25, -0.25]
+
+        """
+        obj = super(GridIterator, cls).__new__(
```
Comment (rhayes777) [inline autoarray/structures/grids.py:930]: super() is fine

Comment (rhayes777) [reply autoarray/structures/grids.py:930]: Oh actually maybe because this is a new function it's not?

File: autoarray/structures/grids.py [hunk 3]
```diff
@@ -781,631 +960,329 @@ def manual_1d(
...
         )
 
-    @property
-    def shape_2d_scaled(self):
-        return (
-            np.amax(self[:, 0]) - np.amin(self[:, 0]),
-            np.amax(self[:, 1]) - np.amin(self[:, 1]),
-        )
+    def interpolated_values_from_values(self, values):
+        return np.einsum("nj,nj->n", np.take(values, self.vtx), self.wts)
```
Comment (rhayes777) [inline autoarray/structures/grids.py:1200]: What does this function do?

File: autoarray/structures/grids.py [hunk 4]
```diff
@@ -781,631 +960,329 @@ def manual_1d(
...
 
-        pixel_scales = grid.mask.pixel_scales
+        a = 0
 
-        pixel_scales = (
-            (grid.shape_2d_scaled[0] + pixel_scales[0]) / (unmasked_sparse_shape[0]),
-            (grid.shape_2d_scaled[1] + pixel_scales[1]) / (unmasked_sparse_shape[1]),
-        )
+        for coords in coordinates:
+            a = a + len(coords)
```
Comment (rhayes777) [inline autoarray/structures/grids.py:1257]: a += len(coords)

File: autoarray/structures/grids.py [hunk 5]
```diff
@@ -1497,54 +1374,337 @@ def squared_distances_from_coordinate(self, coordinate=(0.0, 0.0)):
...
+        obj.pixel_neighbors_size = pixel_neighbors_size.astype("int")
+        obj.nearest_pixelization_1d_index_for_mask_1d_index = (
+            nearest_pixelization_1d_index_for_mask_1d_index
+        )
+
+        return obj
+
+    def __array_finalize__(self, obj):
+
+        if hasattr(obj, "pixel_neighbors"):
```
Comment (rhayes777) [inline autoarray/structures/grids.py:1646]: you could do this as a loop over a tuple of strings with try setattr getattr except AttributeError

File: autoarray/structures/grids.py [hunk 6]
```diff
@@ -1555,7 +1715,172 @@ def extent(self):
...
+
+        unmasked_sparse_for_sparse = sparse_util.unmasked_sparse_for_sparse_from_mask_2d_and_pixel_centres(
+            total_sparse_pixels=total_sparse_pixels,
+            mask_2d=grid.mask,
+            unmasked_sparse_grid_pixel_centres=unmasked_sparse_grid_pixel_centres,
+        ).astype(
+            "int"
+        )
+
+        regular_to_unmasked_sparse = grid_util.grid_pixel_indexes_1d_from_grid_scaled_1d_shape_2d_and_pixel_scales(
```
Comment (rhayes777) [inline autoarray/structures/grids.py:1811]: These function names are a too long.

You can actually require usage of keyword arguments thereby avoiding the need for the "from" part of the function

```python
def grid_pixel_indexes_1d(
    *,
    grid_scaled_1d=...
)
```

Of course this only works if you don't get a clash in function names

File: autoarray/structures/grids.py [hunk 7]
```diff
@@ -1555,7 +1715,172 @@ def extent(self):
...
+            sparse_1d_index_for_mask_1d_index=kmeans.labels_.astype("int"),
+        )
+
+    @property
+    def total_sparse_pixels(self):
+        return len(self.sparse)
+
+
+class GridTransformed(Grid):
+
```
Comment (rhayes777) [inline autoarray/structures/grids.py:1874]: A very useful class

File: autoarray/structures/grids.py [hunk 8]
```diff
@@ -1742,7 +2025,9 @@ def wrapper(profile, grid, *args, **kwargs):
             A grid_like object whose coordinates may be transformed.
         """
 
-        if not isinstance(grid, TransformedGrid) and not isinstance(grid, TransformedGridNumpy):
+        if not isinstance(grid, GridTransformed) and not isinstance(
```
Comment (rhayes777) [inline autoarray/structures/grids.py:2028]:
```python
if not isinstance(grid, (GridTransformed, GridTransformedNumpy))
```

File: autoarray/structures/grids.py [hunk 9]
```diff
@@ -1838,3 +2123,65 @@ def wrapper(profile, grid, *args, **kwargs):
         return func(profile, grid, *args, **kwargs)
 
     return wrapper
+
+
+class MaskedGrid(AbstractGrid):
+    @classmethod
+    def manual_1d(cls, grid, mask, store_in_1d=True):
+
+        if type(grid) is list:
```
Comment (rhayes777) [inline autoarray/structures/grids.py:2132]: prefer isinstance (unless you have some very specific reason here which I guess you might)

File: autoarray/structures/grids.py [hunk 10]
```diff
@@ -1838,3 +2123,65 @@ def wrapper(profile, grid, *args, **kwargs):
...
+
+        if grid.shape[0] != mask.sub_pixels_in_mask:
+            raise exc.GridException(
+                "The input 1D grid does not have the same number of entries as sub-pixels in"
+                "the mask."
+            )
+
+        if store_in_1d:
+            return mask.mapping.grid_stored_1d_from_sub_grid_1d(sub_grid_1d=grid)
+        else:
```
Comment (rhayes777) [inline autoarray/structures/grids.py:2143]: No need for else when there's a return in the if

File: autoarray/structures/grids.py [hunk 11]
```diff
@@ -1838,3 +2123,65 @@ def wrapper(profile, grid, *args, **kwargs):
...
+            return mask.mapping.grid_stored_2d_from_sub_grid_1d(sub_grid_1d=grid)
+
+    @classmethod
+    def manual_2d(cls, grid, mask, store_in_1d=True):
```
Comment (rhayes777) [inline autoarray/structures/grids.py:2147]: Previous comments apply to this fucntion

File: autoarray/structures/grids.py [hunk 12]
```diff
@@ -1838,3 +2123,65 @@ def wrapper(profile, grid, *args, **kwargs):
...
+        sub_grid_1d = grid_util.grid_1d_via_mask_2d(
+            mask_2d=mask,
+            pixel_scales=mask.pixel_scales,
+            sub_size=mask.sub_size,
+            origin=mask.origin,
+        )
+
+        if store_in_1d:
+            return mask.mapping.grid_stored_1d_from_sub_grid_1d(sub_grid_1d=sub_grid_1d)
+        else:
```
Comment (rhayes777) [inline autoarray/structures/grids.py:2186]: Here too

File: test_autoarray/mock/mock_grids.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def float_values_from_grid(profile, grid):
+
+    sersic_constant = (
+        (2 * 2.0)
+        - (1.0 / 3.0)
+        + (4.0 / (405.0 * 2.0))
+        + (46.0 / (25515.0 * 2.0 ** 2))
+        + (131.0 / (1148175.0 * 2.0 ** 3))
```
Comment (rhayes777) [inline test_autoarray/mock/mock_grids.py:17]: What's this all about?

File: test_autoarray/mock/mock_grids.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+                np.add(np.power(np.divide(grid_radii, 0.2), 1.0 / 2.0), -1),
+            )
+        )
+
+    # @grids.grid_like_to_structure
+    # def tuple_values_from_grid(self, grid):
+    #     return np.multiply(2.0, grid)
+    #
+    # @grids.grid_like_to_structure
+    # def float_values_from_grid_returns_list(self, grid):
```
Comment (rhayes777) [inline test_autoarray/mock/mock_grids.py:64]: Delete commented code

File: autoarray/structures/grids.py (hunk 1 above)
Comment (rhayes777) [reply autoarray/structures/grids.py:767]: Given that it's multiple possible types you can use typing.Union

File: autoarray/structures/grids.py (hunk 7 above)
Comment (Jammy2211) [reply autoarray/structures/grids.py:1874]: There was a nasty bug where grids were transformed multiple time, as the grid_like decorator was losing their TrasnformedGrid class.

This is a hacky fix but a good solution needs a refactor of mask.py, which is gonna be a huge undertaking.

File: autoarray/structures/grids.py (hunk 1 above)
Comment (Jammy2211) [reply autoarray/structures/grids.py:767]: Donoe.

File: autoarray/structures/grids.py (hunk 6 above)
Comment (Jammy2211) [reply autoarray/structures/grids.py:1811]: Adpoted this convention, but use of * is restricted due to numba and values with non default values.

File: autoarray/structures/grids.py (hunk 9 above)
Comment (Jammy2211) [reply autoarray/structures/grids.py:2132]: Done

**Pull request merged**